pkgs/development/compilers/openjdk/generic.nix

```diff
@@ -0,0 +1,664 @@
+
+  pkg-config,
+  autoconf,
+  lndir,
+  unzip,
+  ensureNewerSourcesForZipFilesHook,
+
+  cpio,
+  file,
+  which,
+  zip,
+  perl,
+  zlib,
+  cups,
+  freetype,
+  harfbuzz,
+  alsa-lib,
+  libjpeg,
+  giflib,
+  libpng,
+  lcms2,
+  libX11,
+  libICE,
+  libXext,
+  libXrender,
+  libXtst,
+  libXt,
+  libXi,
+  libXinerama,
+  libXcursor,
+  libXrandr,
+  fontconfig,
+
+  setJavaClassPath,
+
+  versionCheckHook,
+
+  bash,
+  liberation_ttf,
+  cacert,
+
+  nixpkgs-openjdk-updater,
+
+  # TODO(@sternenseemann): gtk3 fails to evaluate in pkgsCross.ghcjs.buildPackages
+  # which should be fixable, this is a no-rebuild workaround for GHC.
+  headless ? lib.versionAtLeast featureVersion "21" && stdenv.targetPlatform.isGhcjs,
+
+  enableJavaFX ? false,
+  openjfx17,
+  openjfx21,
+  openjfx23,
+  openjfx ?
+    {
+      "17" = openjfx17;
+      "21" = openjfx21;
+      "23" = openjfx23;
+    }
+    .${featureVersion} or (throw "JavaFX is not supported on OpenJDK ${featureVersion}"),
+
+  enableGtk ? true,
+  gtk3,
+  gtk2,
+  glib,
+
+  temurin-bin-8,
+  temurin-bin-11,
+  temurin-bin-17,
+  temurin-bin-21,
+  temurin-bin-23,
+  jdk-bootstrap ?
+    {
+      "8" = temurin-bin-8;
+      "11" = temurin-bin-11;
+      "17" = temurin-bin-17;
+      "21" = temurin-bin-21;
+      "23" = temurin-bin-23;
+    }
+    .${featureVersion}.override
+      {
+        # when building a headless jdk, also bootstrap it with a headless jdk
+        gtkSupport = !headless;
+      },
+}:
+
+let
+  sourceFile = ./. + "/${featureVersion}/source.json";
+  source = nixpkgs-openjdk-updater.openjdkSource {
+    inherit sourceFile;
+    featureVersionPrefix = tagPrefix + featureVersion;
+  };
+
+  atLeast11 = lib.versionAtLeast featureVersion "11";
+  atLeast17 = lib.versionAtLeast featureVersion "17";
+  atLeast21 = lib.versionAtLeast featureVersion "21";
+  atLeast23 = lib.versionAtLeast featureVersion "23";
+
+  tagPrefix = if atLeast11 then "jdk-" else "jdk";
+  # TODO: Merge these `lib.removePrefix` calls once update scripts have
+  # been run.
+  version = lib.removePrefix tagPrefix (lib.removePrefix "refs/tags/" source.src.rev);
+  versionSplit =
+    # TODO: Remove `-ga` logic once update scripts have been run.
+    builtins.match (if atLeast11 then "(.+)[-+](.+)" else "(.+)-b?(.+)") version;
+  versionBuild = lib.elemAt versionSplit 1;
+
+  # The JRE 8 libraries are in directories that depend on the CPU.
+  architecture =
+    if atLeast11 then
+      ""
+    else
+      {
+        i686-linux = "i386";
+        x86_64-linux = "amd64";
+        aarch64-linux = "aarch64";
+        powerpc64le-linux = "ppc64le";
+      }
+      .${stdenv.system} or (throw "Unsupported platform ${stdenv.system}");
+
+  jdk-bootstrap' = jdk-bootstrap.override {
+    # when building a headless jdk, also bootstrap it with a headless jdk
+    gtkSupport = !headless;
+  };
+in
+
+assert lib.assertMsg (lib.pathExists sourceFile)
+  "OpenJDK ${featureVersion} is not a supported version";
```

Comment:
Lazy evaluation means that the `assert` is the first thing to be evaluated:

```
emily@yuyuko ~/D/nixpkgs ((afb8cd5d))> nix repl -f .
Nix 2.24.9
Type :? for help.
Loading installable ''...
Added 22519 variables.
nix-repl> :b callPackage ./pkgs/development/compilers/openjdk/generic.nix { featureVersion = "123"; } 
error:
       … while calling a functor (an attribute set with a '__functor' attribute)
         at /Users/emily/Developer/nixpkgs/lib/customisation.nix:264:13:
          263|     in if missingArgs == {}
          264|        then makeOverridable f allArgs
             |             ^
          265|        # This needs to be an abort so it can't be caught with `builtins.tryEval`,

       … while evaluating a branch condition
         at /Users/emily/Developer/nixpkgs/lib/customisation.nix:148:7:
          147|     in
          148|       if isAttrs result then
             |       ^
          149|         result // {

       (stack trace truncated; use '--show-trace' to show the full, detailed trace)

       error: OpenJDK 123 is not a supported version
```